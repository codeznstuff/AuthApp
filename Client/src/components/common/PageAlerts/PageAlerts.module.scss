.wrapper {
  width: 100%;
  display: flex;
  justify-content: center;
}
.alertsColumn {
  position: absolute;
  z-index: 100;
  padding-top: 12px;
  width: 500px;
  display: flex;
  flex-direction: column;
}
