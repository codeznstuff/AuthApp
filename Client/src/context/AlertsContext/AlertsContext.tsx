import React, { createContext, useContext, useReducer, useCallback } from 'react';

export interface IAlert {
  variant: 'primary' | 'secondary' | 'success' | 'danger' | 'warning' | 'info' | 'dark' | 'light';
  heading?: string;
  text: string;
  id?: string;
}

interface IAlertAction {
  type: string;
  alert?: IAlert;
}

export const AlertsContext = createContext([]);

export const alertReducer = (state: Array<IAlert>, action: IAlertAction) => {
  switch (action.type) {
    case 'addAlert':
      return [...state, action.alert];
    case 'removeAlert': {
      const newState: Array<IAlert> = [...state];
      newState.shift();
      return newState;
    }
    default:
      return state;
  }
};

export const AlertsProvider = (props: any): any => {
  const [alerts, dispatch] = useReducer(alertReducer, []);
  const ALERT_TIMEOUT = 5000;

  const setRemoveTimeout = () => {
    setTimeout(() => {
      dispatch({ type: 'removeAlert' });
    }, ALERT_TIMEOUT);
  };

  const createAlert = useCallback((alert: IAlert) => {
    alert.id = `${Date.now()}-${alert.text}`;
    dispatch({ type: 'addAlert', alert });
    setRemoveTimeout();
  }, []);

  const context = {
    alerts,
    createAlert
  };

  return <AlertsContext.Provider value={context} {...props} />;
};

export const useAlerts = () => {
  const context: any = useContext(AlertsContext);
  if (context === undefined) {
    throw new Error(`useAlerts must be used within a AlertsProvider`);
  }
  return context;
};
