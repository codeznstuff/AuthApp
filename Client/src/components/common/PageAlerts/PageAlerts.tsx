import React from 'react';
import { Alert } from 'react-bootstrap';
import styles from './PageAlerts.module.scss';
import { useAlerts, IAlert } from '../../../context/AlertsContext/AlertsContext';

const PageAlerts = () => {
  const { alerts } = useAlerts();

  if (!alerts || !alerts.length) return null;

  return (
    <div className={styles.wrapper}>
      <div className={styles.alertsColumn}>
        {alerts.map((alert: IAlert) => {
          return (
            <Alert show variant={alert.variant} key={alert.id}>
              {alert.heading && <Alert.Heading>{alert.heading}</Alert.Heading>}
              {alert.text}
            </Alert>
          );
        })}
      </div>
    </div>
  );
};

export default PageAlerts;
