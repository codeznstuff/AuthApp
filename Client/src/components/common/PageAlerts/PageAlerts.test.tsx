import React from 'react';
import { render, RenderResult, cleanup } from '@testing-library/react';
import PageAlerts from './PageAlerts';
import { AlertsContext, IAlert } from '../../../context/AlertsContext/AlertsContext';

describe.only('PageAlerts', () => {
  let pageAlerts: RenderResult;
  afterEach(cleanup);

  describe('when alerts exist', () => {
    beforeEach(() => {
      const alerts: Array<IAlert> = [{ variant: 'success', text: 'success', id: 1 }, { variant: 'warning', text: 'warning', id: 2 }];
      const context: any = {
        alerts,
        createAlert: () => {}
      };

      pageAlerts = render(
        <AlertsContext.Provider value={context}>
          <PageAlerts />
        </AlertsContext.Provider>
      );
    });

    it('matches snapshot', () => {
      expect(pageAlerts.container.firstChild).toMatchSnapshot();
    });
  });

  describe('when alerts do not exist', () => {
    beforeEach(() => {
      const alerts: Array<IAlert> = [];
      const context: any = {
        alerts,
        createAlert: () => {}
      };

      pageAlerts = render(
        <AlertsContext.Provider value={context}>
          <PageAlerts />
        </AlertsContext.Provider>
      );
    });

    it('matches snapshot', () => {
      expect(pageAlerts.container.firstChild).toMatchSnapshot();
    });
  });
});
