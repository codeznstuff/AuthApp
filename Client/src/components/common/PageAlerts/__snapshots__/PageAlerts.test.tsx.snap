// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageAlerts when alerts do not exist matches snapshot 1`] = `null`;

exports[`PageAlerts when alerts exist matches snapshot 1`] = `
<div
  class="wrapper"
>
  <div
    class="alertsColumn"
  >
    <div
      class="fade alert alert-success show"
      role="alert"
    >
      success
    </div>
    <div
      class="fade alert alert-warning show"
      role="alert"
    >
      warning
    </div>
  </div>
</div>
`;
